R/chat.R

```diff
@@ -121,56 +121,25 @@ Chat <- R6::R6Class(
       invisible(self)
     },
 
-    #' @description A data frame with a `tokens` column that provides the
-    #'   number of input tokens used by user turns and the number of
-    #'   output tokens used by assistant turns.
-    #' @param include_system_prompt Whether to include the system prompt in
-    #'   the turns (if any exists).
-    get_tokens = function(include_system_prompt = FALSE) {
-      turns <- self$get_turns()
-      assistant_turns <- keep(turns, function(x) x@role == "assistant")
-
-      n <- length(assistant_turns)
-      tokens_acc <- map_tokens(assistant_turns, \(turn) turn@tokens)
-      # Combine counts for input tokens (cached and uncached)
-      tokens_acc[, 1] <- tokens_acc[, 1] + tokens_acc[, 3]
-      # Then drop cached tokens counts
-      tokens_acc <- tokens_acc[, 1:2]
-
-      tokens <- tokens_acc
-      if (n > 1) {
-        # Compute just the new tokens
-        tokens[-1, 1] <- tokens[seq(2, n), 1] -
-          (tokens[seq(1, n - 1), 1] + tokens[seq(1, n - 1), 2])
-      }
-
-      # collapse into a single vector
-      tokens_v <- c(t(tokens))
-      tokens_acc_v <- c(t(tokens_acc))
-
-      tokens_df <- data.frame(
-        role = rep(c("user", "assistant"), times = n),
-        tokens = tokens_v,
-        tokens_total = tokens_acc_v,
-        contents = map_chr(turns, turn_contents_preview)
-      )
-
-      if (include_system_prompt && private$has_system_prompt()) {
-        system_turn <- private$.turns[[1]]
-
-        # How do we compute this?
-        tokens_df <- rbind(
-          data.frame(
-            role = "system",
-            tokens = 0,
-            tokens_total = 0,
-            contents = turn_contents_preview(system_turn)
-          ),
-          tokens_df
+    #' @description A data frame with token usage and cost data. There are four
+    #'   columns: `input`, `output`, `cached_input`, and `cost`. There is one
+    #'   row for each pair of user and assistant turns, because token counts
+    #'   and costs are only available after the assistant computes the output.
```

Comment:
```suggestion
    #' @description A data frame with token usage and cost data. There are four
    #'   columns: `input`, `output`, `cached_input`, and `cost`. There is one
    #'   row for each assistant turn, because token counts and costs are only
    #'   available when the API returns the assistant's response.
```

We know I lean toward not enforcing paired user/assistant turns; putting that aside I think it's clearer to just say that we get token counts and costs when we get an assistant response because that's when the actual interaction with the API happens.

Reply:
Yeah, that's much clearer. Thanks!